load-gen/src/main/java/io/smartcat/data/loader/LoadGenerator.java

```diff
@@ -1,155 +1,105 @@
+                previous += calculateConsumedTime(normalizedRate, rate);
+            }
+            for (int i = 0; i < normalizedRate; i++) {
+                if (!dataSource.hasNext(elapsed)) {
+                    LOGGER.info("Reached end of data source. Stopping load generator.");
+                    terminate.set(true);
+                    break infiniteWhile;
+                }
+                T data = dataSource.getNext(fromBeginning);
+                worker.accept(data);
+            }
         }
+        dataSource.terminate();
+        rateGenerator.terminate();
+        worker.terminate();
     }
 
     /**
-     * LoadGenerator builder class.
+     * Stops load generator.
      */
-    public static class Builder {
-
-        private int targetRate = 1000;
-        private boolean collectMetrics = true;
-        private WorkTask workTask = new NoOpWorkTask();
-        private DataSource dataSource = new NoOpDataSource();
-        private RefillStrategy refillStrategy;
-        private SleepStrategy sleepStrategy = SleepStrategies.nanosecondSleepStrategy(1);
-
-        /**
-         * Set target rate per second.
-         *
-         * @param targetRate target rate
-         * @return builder
-         */
-        public Builder withTargetRate(int targetRate) {
-            this.targetRate = targetRate;
-            return this;
-        }
-
-        /**
-         * Set this to true if metrics should be collected.
-         *
-         * @param collectMetrics collect metrics
-         * @return builder
-         */
-        public Builder withCollectMetrics(boolean collectMetrics) {
-            this.collectMetrics = collectMetrics;
-            return this;
-        }
-
-        /**
-         * WorkTask interface implementation. This is what load generator executes at a given rate.
-         *
-         * @param workTask work task implementation
-         * @return builder
-         */
-        public Builder withWorkTask(WorkTask workTask) {
-            this.workTask = workTask;
-            return this;
-        }
-
-        /**
-         * DataSource interface implementation. Provides data source for all executed tasks.
-         *
-         * @param dataSource data source implementation
-         * @return builder
-         */
-        public Builder withDataSource(DataSource dataSource) {
-            this.dataSource = dataSource;
-            return this;
-        }
-
-        /**
-         * Refill strategy implementation for generating token bucket tokens.
-         * Default implementation is {@code io.smartcat.data.loader.tokenbuket.FixedRateRefillStrategy}.
-         *
-         * @param refillStrategy refill strategy implementation
-         * @return builder
-         */
-        public Builder withRefillStrategy(RefillStrategy refillStrategy) {
-            this.refillStrategy = refillStrategy;
-            return this;
-        }
+    public void stop() {
+        terminate.set(true);
```

Comment:
Few logging messages at the INFO level (here and there) would make it more user-friendlier in CLI.